go/libraries/doltcore/table/editor/table_edit_accumulator.go

```diff
@@ -0,0 +1,341 @@
+// When a rollback occurs uncommitted changes are dropped.
+//
+// In addition to the in memory edits, the changes are applied to committedEA when a commit occurs. It is possible
+// for the uncommitted changes to become so large that they need to be flushed to disk.  When this happens we track the
+// in memory changes in uncommittedFlushed. When materializing the map, edits must be applied to the row data in order
+// so the types.EditAccumulators are numbered so their changes can be applied in order.
+type tableEditAccumulatorImpl struct {
+	nbf *types.NomsBinFormat
+
+	// initial state of the map
+	rowData types.Map
+
+	// in memory changes which will be applied to the
+	committed          *inMemModifications
+	uncommitted        *inMemModifications
+	uncommittedFlushed *inMemModifications
+
+	// accumulatorIdx defines the order in which types.EditAccumulators will be applied
+	accumulatorIdx uint64
+
+	// flusher manages flushing of the types.EditAccumulators to disk when needed
+	flusher *edits.DiskEditFlusher
+
+	// committedEaIds tracks ids of edit accumulators which have changes that have been committed
+	committedEaIds *set.Uint64Set
+	// uncommittedEAIds tracks ids of edit accumulators which have not been committed yet.
+	uncommittedEaIds *set.Uint64Set
+
+	// commitEA is the types.EditAccumulator containing the committed changes that are being accumulated currently
+	commitEA types.EditAccumulator
+	// commitEAId is the id used for ordering the commitEA with other types.EditAccumulators that will be applied when
+	// materializing all changes.
+	commitEAId uint64
+}
+
+// Get returns a *doltKVP if the current TableEditAccumulator contains the given key, or it exists in the row data.
+// This assumes that the given hash is for the given key.
+func (tea *tableEditAccumulatorImpl) Get(ctx context.Context, keyHash hash.Hash, key types.Tuple) (*doltKVP, bool, error) {
+	// in order of most recent changes to least recent falling back to whats in the materialized row data
+	orderedMods := []*inMemModifications{tea.uncommitted, tea.uncommittedFlushed, tea.committed}
+	for _, mods := range orderedMods {
+		kvp, added, deleted := mods.Get(keyHash)
+
+		if added {
+			return kvp, true, nil
+		} else if deleted {
+			return nil, false, nil
+		}
+	}
+
+	v, ok, err := tea.rowData.MaybeGetTuple(ctx, key)
+	if err != nil {
+		return nil, false, err
+	}
+	if !ok {
+		return nil, false, nil
+	}
+
+	return &doltKVP{k: key, v: v}, true, err
+}
+
+func (tea *tableEditAccumulatorImpl) flushUncommitted() {
+	ea := edits.NewAsyncSortedEditsWithDefaults(tea.nbf)
+	for _, kvp := range tea.uncommitted.adds {
+		ea.AddEdit(kvp.k, kvp.v)
+	}
+
+	for _, key := range tea.uncommitted.deletes {
+		ea.AddEdit(key, nil)
+	}
+
+	tea.uncommittedEaIds.Add(tea.accumulatorIdx)
+	tea.flusher.Flush(ea, tea.accumulatorIdx)
+	tea.accumulatorIdx++
+
+	tea.uncommittedFlushed.MergeIn(tea.uncommitted)
+	tea.uncommitted = NewInMemModifications()
+}
```

Comment:
It looks like we're now handling the auto-flushing behavior on the same thread that writes are being done, where as previously it was asynchronous. I decided to run `master` against this branch on a 1-million-row SQL file and:

```
+---------+-----------------+
| master  | bh/tea-refactor |
+---------+-----------------+
| 27.843s | 34.109s         |
+---------+-----------------+
```

Takes 22% longer so it's not a massive difference, but it's primarily masked by how slow our SQL analyzer is, so I'm sure that as we optimize the analyzer, this gap will become larger and larger. I remember way back when making the first iteration of the `TableEditAccumulator`, the SQL path showed a 5% increase where as the import path was like 70-80% because there was just so much analyzer overhead in the SQL path.

Do you have any near-at-hand ideas on how to make this asynchronous? If not it's fine for now, but we may be optimizing this later if not.

Reply:
so the flushes happen asynchronously in the `tea.flusher.Flush(ea, tea.accumulatorIdx)` call.  I played with your test for a bit.  We are 20% slower in your case.  This only happens when the inserts are already ordered, the number of inserts isn't massive, and the number of columns isn't huge.  If any of these isn't true the perf is better and there is less garbage to collect. Been playing with other things to make it even better but haven't found a way to improve this special case.